### PR TITLE
Redact secrets in Misk config admin dashboard

### PR DESCRIPTION
**Pull request by adrw**
Misk has since 2018 been missing Secret class redaction for the Config tab, which resulted in the Config tab being disabled in real environments given secrets leaking concerns. This change adds proper redaction of the class and makes it safe to include by default going forward. 

Notably any Secret within a service config will be redacted. Non-Secret fields that should be redacted can be using the `@Redact` annotation on the data class or property.

Screenshot 2023-03-14 at 16 57 11

- [x] Add tests
- [x] Get review from security
- [x] Add tests that new redaction annotations don't change access within code

### CODE REVIEW

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (cconroy) [inline misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:59]: (existing, but shouldn't this be a forEach?)

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (cconroy) [inline misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:58]: i would expect the unsafe `toYaml` rather than `toRedactedYaml` here?

File: misk/src/main/kotlin/misk/config/MiskConfig.kt
Comment (cconroy) [inline misk/src/main/kotlin/misk/config/MiskConfig.kt:208]: I think this comment applies to both branches of this conditional?

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:58]: I renamed the `toYaml` function to `toRedactedYaml`

Comment (cconroy) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:61]: I think we should simplify the safe/unsafe logic. Something like:

```
  private fun generateConfigResources(
    appName: String,
    deployment: Deployment,
    config: Config
  ): Map<String, String?> {
    val rawYamlFiles = MiskConfig.loadConfigYamlMap(appName, deployment, listOf())
    val configFileContents = linkedMapOf<String, String?>()

    configFileContents.put("Effective Config", MiskConfig.toRedactedYaml(config, ResourceLoader.SYSTEM))
    if (mode == ConfigTabMode.UNSAFE_LEAK_MISK_SECRETS) {
      rawYamlFiles.forEach { configFileContents.put(it.key, it.value) }
    }

    configFileContents.put("JVM", jvmMetadataAction.getRuntime())
    return configFileContents
  }
```

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (cconroy) [inline misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:34]: I'm not clear what "Need to figure out how to get the overrides" indicates here. Do we think this is still relevant?

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (cconroy) [inline misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:36]: I know the shape of this pattern is existing, but it has effectively not been used for a while.

I think it's generally good idea to have unit tests specifically for regexes to verify they are behaving as we expect. At first glance, it would seem this matcher would be overly greedy if we had a key inside the value like

```
  some_key: "The quick brown fox found his password: it was 'what does the fox say?'"
```

becomes 

```
  some_key: "The quick brown fox found his password: ████████"
```

Though, I'd be more concerned about _missing_ something than being overly greedy for an edge case like that...

YAML does have [a lot of ways](https://stackoverflow.com/a/21699210) to write multi-line values which I think will defeat this pattern and evade redaction. Though, maybe not a concern in this flow if a trip through deserialization normalizes those to always work with this pattern?

Though, I think the premise of this whole key matching is a bit flawed...Anything that you include directly in your config _should not be a secret_ because it's checked in to source control, anybody in the company can download the deployable artifact, etc. And, sure enough, in general we see things like `passphrase: doesntmatter` in our configs.

Using regex here ultimately feels like the wrong layer to do what we need. Possibly worth thinking about redaction from the perspective of keys that _load_ their content from the filesystem/classpath and do the scrubbing a layer _above_ serialization?

And all of the above ^ has me continuing to wonder if we wouldn't just be better off biting the bullet of introducing a `!secret` directive in misk as well. Though I don't want to hold up useful config to wait for all existing secrets to move to a new structure like that...

Happy to chat/pair through some of this stuff f2f tomorrow.

File: samples/exemplar/src/main/resources/exemplar-common.yaml
Comment (adrw) [reply samples/exemplar/src/main/resources/exemplar-common.yaml:8]: That was to support booting the exemplar service locally to confirm that it was redacted (ie. screenshot in the PR)

File: misk-admin/src/test/kotlin/misk/web/metadata/config/ConfigMetadataActionTest.kt
Comment (adrw) [reply misk-admin/src/test/kotlin/misk/web/metadata/config/ConfigMetadataActionTest.kt:143]: Unfortunate side effect of the Jackson redaction serializer which redacts the contents but still outputs valid YAML with the quotes.

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:34]: Good call, looks like that still is a valid TODO based on further context https://github.com/cashapp/misk/commit/7619df2889585ad85dabeefbe25f3f676153d826

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt]: Thanks for the fuller thoughts there, I managed to get annotated redaction working instead which I think covers the effective config case sufficiently. Let me know if there's any case that we might need additional coverage for.

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt]: Managed to get it working!

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt]: I ended up getting annotated redaction working so I removed the YAML RegEX redaction to simplify the security model and implementation.

**Comment by AlexSzlavik** [comment]
Am I understanding correctly that you've changed direction now and are using an annotation to opt into redaction of fields as oppose to applying a regex match to yaml keys? 
I like the approach of building the primitives to perform redaction more effectively.

**Review by AlexSzlavik — APPROVED** [review]
I like this new approach

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (cconroy) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:49]: The only use of this is to display to service owners in their admin dashboard...Since we wouldn't ever run the unsafe mode in a real production environment, I would say we should just remove any JVM metadata we don't want exposed in the admin console/endpoint.

Though, I would think our admin console ACLs should be sufficient mitigation for this concern. Anyone who is ACL'd to the admin console is an insider and could therefore determine this information out of band if they really wanted to.

File: misk/src/main/kotlin/misk/config/MiskConfig.kt
Comment (adrw) [reply misk/src/main/kotlin/misk/config/MiskConfig.kt:391]: Yeah I was hoping to do that but it looks like by this point it's already been resolved to the value.

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:49]: The JVM config was present in both safe/unsafe (see L53 and L58 of old file). I agree with @cconroy that the admin ACL is probably sufficient protection. I'm open to redacting any fields in JVM that you'd like to redact from engineers.

File: misk/src/main/kotlin/misk/config/MiskConfig.kt
Comment (adrw) [reply misk/src/main/kotlin/misk/config/MiskConfig.kt]: I think I found a way to do it, see the updated screenshot. If tests go green we should be good.

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (AlexSzlavik) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt:49]: > Anyone who is ACL'd to the admin console is an insider and could therefore determine this information out of band if they really wanted to.
We've seen at least 1 incident where admin dashboards were opened to the public web. Luckily it was for a benign service. It made me think of defense in depth here. Not a hard requirement for me

File: misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt
Comment (adrw) [reply misk-admin/src/main/kotlin/misk/web/metadata/config/ConfigMetadataAction.kt]: Going to pull this offline so I can go through some practical examples, but for effective config I agree it's not necessary. For the unsafe show all the raw YAMLs mode we probably want to add support eventually.

**Comment by adrw** [comment]
@cconroy want to do a final review? I think it's ready to merge and test with a few services before eventually changing the default config tab mode to `SHOW_REDACTED_EFFECTIVE_CONFIG`.

File: misk/src/main/kotlin/misk/config/MiskConfig.kt
Comment (cconroy) [inline misk/src/main/kotlin/misk/config/MiskConfig.kt:392]: 🔥

**Review by cconroy — APPROVED** [review]

**Pull request merged**
